src/nvim/mark.c

```diff
@@ -887,6 +887,18 @@ void ex_changes(exarg_T *eap)
  */
 void mark_adjust(linenr_T line1, linenr_T line2, long amount, long amount_after)
 {
+  mark_adjust_internal(line1, line2, amount, amount_after, true);
+}
+
+void mark_adjust_nofold(linenr_T line1, linenr_T line2, long amount,
```

Comment:
`mark_adjust_nofold()` does work in the same way as `mark_adjust()` (apart from needing updates of the folds separately).

The way I read that bit of the documentation is:
 when you're calling `changed_*() ...`, make sure to call `mark_adjust()` first.
Not 
make sure to call one of `changed_*() ...` after each call to `mark_adjust()`.

Calling `mark_adjust()` more than once before calling one of those mentioned functions is also done in `do_filter()` -- the same place that I think there's a bug, but not because of that.

Agreed on the documentation about the relationship with folds, but just by way of explanation: there's no special relationship between `mark_adjust_nofold()` and `foldSwapRange()`.
`foldSwapRange()` was just a function to do the specific moving around of folds that I needed in this case.